roles/openshift_grafana/tasks/main.yml

```diff
@@ -2,121 +2,142 @@
+- name: Add view role to grafana user
   oc_adm_policy_user:
-    user: grafana-ocp
+    user: "{{ grafana_user }}"
     resource_kind: cluster-role
     resource_name: view
     state: present
-    role_namespace: "{{ gf_prometheus_namespace }}"
-
-- name: Get grafana route
-  oc_obj:
-    kind: route
-    name: grafana
-    namespace: grafana
-  register: route
-
-- name: Get prometheus route
-  oc_obj:
-    kind: route
-    name: prometheus
-    namespace: "{{ gf_prometheus_namespace }}"
-  register: route
+    role_namespace: "{{ grafana_prometheus_namespace }}"
 
 - name: Get the prometheus SA
-  oc_serviceaccount_secret:
-    state: list
-    service_account: prometheus
-    namespace: "{{ gf_prometheus_namespace }}"
+  shell: oc sa get-token {{ grafana_prometheus_sa }} -n {{ grafana_prometheus_namespace }}
   register: sa
 
-- name: Get the management SA bearer token
-  set_fact:
-    management_token: "{{ sa.results | oo_filter_sa_secrets }}"
-
-- name: Ensure the SA bearer token value is read
-  oc_secret:
-    state: list
-    name: "{{ management_token }}"
-    namespace: "{{ gf_prometheus_namespace }}"
-  no_log: True
-  register: sa_secret
-
-- name: Get the SA bearer token for prometheus
-  set_fact:
-    token: "{{ sa_secret.results.encoded.token }}"
+- name: Get grafana route
+  shell: oc get route grafana -o jsonpath='{.spec.host}' -n {{ grafana_namespace }}
+  register: grafana_route
 
-- name: Convert to json
-  var:
-    ds_json: "{{ gf_body_tmp }} | to_json }}"
+- name: Get prometheus route
+  shell: oc get route prometheus -n {{ grafana_prometheus_namespace }} -o jsonpath='{.spec.host}'
+  register: grafana_prom
 
-- name: Set protocol type
-  var:
-    protocol: "{{ 'https' if {{ gf_oauth }} == true else 'http' }}"
+- set_fact: ds_json={{ datasource_payload | to_json }}
+- set_fact: prometheus={{ grafana_prom.stdout }}
+- set_fact: route={{ grafana_route.stdout }}
+- set_fact: token={{ sa.stdout }}
+- set_fact:
+        payload_data: "{{ ds_json | regex_replace('grafana_name', grafana_datasource_name ) | regex_replace('prometheus_url', prometheus ) | regex_replace('satoken',  token ) }}"
 
-- name: Add gf datasrouce
+- name: Add new datasrouce to grafana
   uri:
-    url: "{{ protocol }}://{{ route }}/api/datasources"
+    url: "https://{{ route }}/api/datasources"
     user: admin
     password: admin
     method: POST
-    body: "{{ ds_json | regex_replace('grafana_name', {{ gf_datasource_name }}) | regex_replace('prometheus_url', 'https://'{{ prometheus }} ) | regex_replace('satoken', {{ token }}) }}"
+    body: "{{ payload_data }}"
+    body_format: json
     headers:
       Content-Type: "Content-Type: application/json"
   register: add_ds
 
-- name: Regex setup ds name
+- name: Pull grafana yaml from openshift/origin
+  get_url:
+    url: "{{ grafana_base_url }}/openshift-cluster-monitoring.json"
+    dest: /tmp/openshift-cluster-monitoring.json
+
+- name: Regex set data soure name for openshift dashboard
```

Comment:
I've recently added support for yedit to modify json.

I'm a little uncomfortable running replace(sed-like) operations on structured data.

Reply:
@mrsiano and I discussed this pattern.   In the future when this role settles into a production use case we will want to templatize the json files.  Currently this role fetches the json files from a remote github location.  I feel this is undesirable as the older versions will get the latest features.  We should join monitoring templates to the version and allow a template to receive updated monitoring via variables.